xbmc/favourites/FavouritesService.cpp

```diff
@@ -0,0 +1,228 @@
+      {
+        const CFileItemPtr item(std::make_shared<CFileItem>(name));
+        item->SetPath(favURL);
+        if (thumb)
+          item->SetArt("thumb", thumb);
+        items.Add(item);
+      }
+    }
+    favourite = favourite->NextSiblingElement("favourite");
+  }
+  return true;
+}
+
+CFavouritesService::CFavouritesService(std::string userDataFolder) : m_userDataFolder(std::move(userDataFolder))
+{
+  CFileItemList items;
+  std::string favourites = "special://xbmc/system/favourites.xml";
+  if(XFILE::CFile::Exists(favourites))
+    LoadFromFile(favourites, m_favourites);
+  else
+    CLog::Log(LOGDEBUG, "CFavourites::Load - no system favourites found, skipping");
+
+  favourites = URIUtils::AddFileToFolder(m_userDataFolder, "favourites.xml");
+  if(XFILE::CFile::Exists(favourites))
+    LoadFromFile(favourites, m_favourites);
+  else
+    CLog::Log(LOGDEBUG, "CFavourites::Load - no userdata favourites found, skipping");
+}
+
+CFavouritesService::~CFavouritesService()
+{
+}
+
+bool CFavouritesService::Persist()
+{
+  CXBMCTinyXML doc;
+  TiXmlElement xmlRootElement("favourites");
+  TiXmlNode *rootNode = doc.InsertEndChild(xmlRootElement);
+  if (!rootNode)
+    return false;
+
+  for (int i = 0; i < m_favourites.Size(); i++)
+  {
+    const CFileItemPtr item = m_favourites[i];
+    TiXmlElement favNode("favourite");
+    favNode.SetAttribute("name", item->GetLabel().c_str());
+    if (item->HasArt("thumb"))
+      favNode.SetAttribute("thumb", item->GetArt("thumb").c_str());
+
+    const CURL url(item->GetPath());
+    TiXmlText execute(CURL::Decode(url.GetHostName()));
+    favNode.InsertEndChild(execute);
+    rootNode->InsertEndChild(favNode);
+  }
+
+  auto path = URIUtils::AddFileToFolder(m_userDataFolder, "favourites.xml");
+  return doc.SaveFile(path);
+}
+
+bool CFavouritesService::Save(const CFileItemList& items)
+{
+  {
+    CSingleLock lock(m_criticalSection);
+    m_favourites.Clear();
+    m_favourites.Copy(items);
+    Persist();
+  }
+  OnUpdated();
+  return true;
+}
+
+void CFavouritesService::OnUpdated()
+{
+  ANNOUNCEMENT::CAnnouncementManager::GetInstance().Announce(ANNOUNCEMENT::GUI, "xbmc", "OnFavouritesUpdated");
+}
+
+bool CFavouritesService::AddOrRemove(const CFileItem& item, int contextWindow)
+{
+  CURL url;
+  url.SetProtocol("favourites");
+  url.SetHostName(CURL::Encode(GetExecutePath(item, contextWindow)));
+
+  CSingleLock lock(m_criticalSection);
+  CFileItemPtr match = m_favourites.Get(url.Get());
+  if (match)
+  { // remove the item
+    m_favourites.Remove(match.get());
+  }
+  else
+  { // create our new favourite item
+    const CFileItemPtr favourite(std::make_shared<CFileItem>(item.GetLabel()));
+    if (item.GetLabel().empty())
+      favourite->SetLabel(CUtil::GetTitleFromPath(item.GetPath(), item.m_bIsFolder));
+    favourite->SetArt("thumb", item.GetArt("thumb"));
+    favourite->SetPath(url.Get());
+    m_favourites.Add(favourite);
+  }
+  Persist();
+  lock.Leave();
```

Comment:
In other places in this file, a scoped <code>CSingleLock</code> instance is used, which I find more elegant and straight forward to read.